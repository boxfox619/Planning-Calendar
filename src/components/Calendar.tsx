import * as React from 'react';
import * as moment from 'moment';
import styled from 'styled-components';
import { CalendarMode } from '../models/CalendarMode';
import { WeeklyCalendar } from '../components/weekCalendar/WeeklyCalenldar';
import { MonthCalendar } from '../components/monthCalendar/MonthCalendar';
import { Task } from '../models';

const Container = styled.div`
    display: flex;
    flex-flow: column;
    padding: 0 10px;
`;
Container.displayName = 'Container';

interface OwnProps {
    currentMoment: moment.Moment,
    mode: CalendarMode,
    tasks: Task[],
    onSelect: (selectedMoment?: moment.Moment | Task) => void,
    onUpdate: (task: Task) => void
}

type Props = OwnProps & React.HTMLAttributes<HTMLDivElement>;

export const Calendar: React.FC<Props> = ({ currentMoment, mode, tasks, onSelect, onUpdate, ...divProps }) => {

    const handleClick = (e: React.MouseEvent<HTMLDivElement, MouseEvent>) => {
        const target = e.target as HTMLElement;
        if (target.dataset.taskid) {
            onSelect(tasks.find(task => task.id === Number(target.dataset.taskid)));
        } else if (target.dataset.datetime) {
            onSelect(moment.parseZone(target.dataset.datetime))
        }
    }
    const handleDragStart = (e: React.DragEvent) => {
        const taskId = (e.target as HTMLElement).dataset.taskid;
        if (taskId) {
            e.dataTransfer.setData("taskId", taskId);
        }
    }
    const handleDrop = (e: React.DragEvent) => {
        const taskId = e.dataTransfer.getData('taskId');
        const datetime = e.dataTransfer.getData('datetime') || (e.target as HTMLElement).dataset.datetime;
        const dropTaskId = (e.target as HTMLElement).dataset.taskid;
        const dropTask = tasks.find(t => t.id === Number(dropTaskId));
        const currentTask = tasks.find(t => t.id === Number(taskId));
        if (!currentTask) {
            return;
        }
        const newTask = { ...currentTask };
        if (datetime) {
            const date = moment.parseZone(datetime).format('YYYY-MM-DD');
            const hour = moment.parseZone(datetime).hour();
            newTask.date = date;
            newTask.endHour += hour - newTask.startHour;
            newTask.startHour = hour;
        } else if (dropTask) {
            newTask.date = dropTask.date;
        }
        onUpdate(newTask);
    };
    const handleDragOver = (e: React.DragEvent) => {
        const target = e.target as HTMLElement;
        if (target.dataset.datetime) {
            e.dataTransfer.setData('datetime', target.dataset.datetime);
        }
    }

    return (
        <Container {...divProps}
            onClick={handleClick}
            onDragStart={handleDragStart}
            onDragOver={handleDragOver}
            onDrop={handleDrop}>
            {mode === CalendarMode.Month && (
                <MonthCalendar
                    style={{ flex: 1 }}
                    currentMoment={currentMoment}
                    tasks={tasks}
                />
            )}
            {mode === CalendarMode.Week && (
                <WeeklyCalendar
                    style={{ flex: 1 }}
                    currentMoment={currentMoment}
                    tasks={tasks}
                />
            )}
        </Container>
    )
}
