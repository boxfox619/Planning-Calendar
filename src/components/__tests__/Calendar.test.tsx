import * as React from 'react';
import * as Enzyme from 'enzyme';
import { Calendar } from '../Calendar';
import { CalendarMode } from '../../models/CalendarMode';
import * as moment from 'moment';
import { Task } from '../../models';

describe('<Calendar />', () => {
    const testMoment = moment.parseZone('2014-02-27T01:00:00Z');
    let component = null;
    const task = new Task(12, '이름', '2019-01-21', 3, 12);
    const tasks = [task];
    const selectMock = jest.fn();
    const updateMock = jest.fn();

    it('renders correctly', () => {
        component = Enzyme.shallow(<Calendar currentMoment={testMoment} mode={CalendarMode.Month} tasks={tasks} onSelect={selectMock} onUpdate={updateMock} />);
    });

    it('should match snapshot', () => {
        expect(component).toMatchSnapshot();
    });

    it('should open dialog', () => {
        expect(component.find('Container').length).toBe(1);
        component.find('Container').simulate('click', { target: { dataset: { taskid: task.id } } };
        expect(selectMock.mock.calls.length).toBe(1);
    });

    it('should handle dragging', () => {
        expect(component.find('Container').length).toBe(1);
        const dataTransferMock = jest.fn();
        const mockEvent = { dataTransfer: { setData: dataTransferMock }, target: { dataset: { taskid: task.id } } };
        component.find('Container').at(0).simulate('dragStart', mockEvent);
        expect(dataTransferMock.mock.calls.length).toBe(1);
        expect(dataTransferMock.mock.calls[0][0]).toBe('taskId');
        expect(dataTransferMock.mock.calls[0][1]).toBe(task.id);
    });

    it('should handle dragover', () => {
        expect(component.find('Container').length).toBe(1);
        const dataTransferMock = jest.fn();
        const mockEvent = { dataTransfer: { setData: dataTransferMock }, target: { dataset: { datetime: testMoment.toISOString() } } };
        component.find('Container').at(0).simulate('dragOver', mockEvent);
        expect(dataTransferMock.mock.calls.length).toBe(1);
        expect(dataTransferMock.mock.calls[0][0]).toBe('datetime');
        expect(dataTransferMock.mock.calls[0][1]).toBe(testMoment.toISOString());
    });

    it('should update task when drop', () => {
        expect(component.find('Container').length).toBe(1);
        const mockEvent = { dataTransfer: { getData: (key: string) => key === 'taskId' ? task.id : testMoment.toISOString() }, target: { dataset: { taskid: task.id } } };
        component.find('Container').at(0).simulate('drop', mockEvent);
        expect(updateMock.mock.calls.length).toBe(1);
        expect(updateMock.mock.calls[0][0].date).toBe('2014-02-27');
    });
});