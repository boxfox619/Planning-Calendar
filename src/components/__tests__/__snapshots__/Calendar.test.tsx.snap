// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Calendar /> should match snapshot 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Unknown
    currentMoment={"2014-02-27T01:00:00.000Z"}
    mode="월"
    onSelect={[Function]}
    onUpdate={[Function]}
    tasks={
        Array [
            Task {
              "date": "2019-01-21",
              "endHour": 12,
              "id": 12,
              "name": "이름",
              "startHour": 3,
            },
          ]
    }
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "function",
    "props": Object {
      "children": Array [
        <Unknown
          currentMoment={"2014-02-27T01:00:00.000Z"}
          style={
                    Object {
                              "flex": 1,
                            }
          }
          tasks={
                    Array [
                              Task {
                                "date": "2019-01-21",
                                "endHour": 12,
                                "id": 12,
                                "name": "이름",
                                "startHour": 3,
                              },
                            ]
          }
/>,
        false,
      ],
      "onClick": [Function],
      "onDragOver": [Function],
      "onDragStart": [Function],
      "onDrop": [Function],
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "currentMoment": "2014-02-27T01:00:00.000Z",
          "style": Object {
            "flex": 1,
          },
          "tasks": Array [
            Task {
              "date": "2019-01-21",
              "endHour": 12,
              "id": 12,
              "name": "이름",
              "startHour": 3,
            },
          ],
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      false,
    ],
    "type": Object {
      "$$typeof": Symbol(react.forward_ref),
      "attrs": Array [],
      "componentStyle": ComponentStyle {
        "componentId": "sc-jTzLTM",
        "isStatic": true,
        "rules": Array [
          "
    display: flex;
    flex-flow: column;
    padding: 0 10px;
",
        ],
      },
      "displayName": "Container",
      "foldedComponentIds": Array [],
      "render": [Function],
      "styledComponentId": "sc-jTzLTM",
      "target": "div",
      "toString": [Function],
      "warnTooManyClasses": [Function],
      "withComponent": [Function],
    },
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {
        "children": Array [
          <Unknown
            currentMoment={"2014-02-27T01:00:00.000Z"}
            style={
                        Object {
                                    "flex": 1,
                                  }
            }
            tasks={
                        Array [
                                    Task {
                                      "date": "2019-01-21",
                                      "endHour": 12,
                                      "id": 12,
                                      "name": "이름",
                                      "startHour": 3,
                                    },
                                  ]
            }
/>,
          false,
        ],
        "onClick": [Function],
        "onDragOver": [Function],
        "onDragStart": [Function],
        "onDrop": [Function],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "currentMoment": "2014-02-27T01:00:00.000Z",
            "style": Object {
              "flex": 1,
            },
            "tasks": Array [
              Task {
                "date": "2019-01-21",
                "endHour": 12,
                "id": 12,
                "name": "이름",
                "startHour": 3,
              },
            ],
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        false,
      ],
      "type": Object {
        "$$typeof": Symbol(react.forward_ref),
        "attrs": Array [],
        "componentStyle": ComponentStyle {
          "componentId": "sc-jTzLTM",
          "isStatic": true,
          "rules": Array [
            "
    display: flex;
    flex-flow: column;
    padding: 0 10px;
",
          ],
        },
        "displayName": "Container",
        "foldedComponentIds": Array [],
        "render": [Function],
        "styledComponentId": "sc-jTzLTM",
        "target": "div",
        "toString": [Function],
        "warnTooManyClasses": [Function],
        "withComponent": [Function],
      },
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromError": true,
          "getDerivedStateFromProps": Object {
            "hasShouldComponentUpdateBug": false,
          },
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
